### PR TITLE
Improved contact form items

### PR DESCRIPTION
**Pull request by thednp**
**Form Item Builder Class**
* the `class-fw-option-type-form-builder-item.php` was breaking the normal structure of shortcode by adding an extra folder `options/options.php` instead of just `options/php`, this was confusing why it didn't work as expected in the first place

**Form header title**
* missing extra options

**Radio**
* moved `layout` option out of class into a new `options.php` file

**Checkboxes**
* moved `layout` option out of class into a new `options.php` file

**Select**
* allow users to customize frontend static, users might use selectize

**reCaptcha**
* added a `static.php` file users can override to customize the front-end scripting

### CODE REVIEW

File: includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php [hunk 1]
```diff
@@ -67,12 +67,12 @@ public function get_value_from_item( $value ) {
 	}
 
 	/**
-	 * Search in '/extensions/forms/{builder_type}/items/{item_type}/options/options.php'
+	 * Search in '/extensions/forms/{builder_type}/items/{item_type}/options.php'
```
Comment (andreiglingeanu) [inline includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php]: Well, that might be a breaking change. @ViorelEremia be careful with that. There may be developers that already have overrided t `options/options.php` and may expect it to be that way.

**Review by andreiglingeanu — APPROVED** [review]

File: includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php (hunk 1 above)
Comment (thednp) [reply includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php]: So far I am the only one [talking about this improvement](https://github.com/ThemeFuse/Unyson/issues/1738), I don't think it's gonna be a problem. And the change to the new version is super easy anyway.

Comment (danyj) [reply includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php]: I am overriding original and I am sure other devs did not see the proposed change, by so many installs by now I would be very careful with this

Comment (thednp) [reply includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php]: So it's OK then, we can skip that change, I was hoping to get that back to "normal". I would have to make some changes in all items (radio, checkboxes) to reflect that.

Comment (danyj) [reply includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php]: Actually im overriding extensions\forms\extensions\contact-forms\shortcodes\contact-form\options.php

still digging trough your changes

**Comment by danyj** [comment]
@thednp  Please explain again what is bad , I cant seem to find the   options/options.php   that you are referring to

**Comment by thednp** [comment]
> @thednp Please explain again what is bad , I cant seem to find the options/options.php that you are referring to

This [change](https://github.com/ThemeFuse/Unyson-Forms-Extension/pull/52/commits/01447cf1a75aeaeefa1fe32eddee14d3f6cceaa4) should correct the incorrect structure and have all [form items](https://github.com/thednp/Unyson-Forms-Extension/tree/bebcdbc07adde56d0bce912fd1d7fd724f1824c9/includes/option-types/form-builder/items) use `checkboxes/options.php` to add additional options instead of the current `checkboxes/options/options.php` framework location.

So far the changes have been applied to `checkboxes`, `radio` and `select` but it can be used for any other form item.

In addition the extra options feature wasn't added to the form title item, just another sweet cake I found and fixed.

**Comment by danyj** [comment]
ok , I see , did not touch that one  https://github.com/ThemeFuse/Unyson-Forms-Extension/blob/master/includes/option-types/form-builder/extends/class-fw-option-type-form-builder-item.php#L75

I did override for views  https://github.com/ThemeFuse/Unyson-Forms-Extension/tree/master/extensions/contact-forms/views

and options 

https://github.com/ThemeFuse/Unyson-Forms-Extension/blob/master/extensions/contact-forms/shortcodes/contact-form/options.php

**Comment by thednp** [comment]
So you would have absolutely no issue with my master version into your setup. That's why I said, I may be the only one who could have been affected by these changes.

**Comment by danyj** [comment]
look ,  to be sure why dont you do a check for old way and if no file is there than use the new way, this way it is backwards compatible and the new way is in

**Comment by danyj** [comment]
and yes you right , the MVC should have been followed   static, view, options

**Comment by thednp** [comment]
@danyj it's not only about that, but it was breaking the pattern we're all used with Unyson over the years.

**Comment by thednp** [comment]
@andreiglingeanu @ViorelEremia one last thing is needed for forms: the ability to override the form errors, as currently there are only [some hacks](https://github.com/ThemeFuse/Unyson/issues/1738#issuecomment-238222718) to help solve this issue.

I would like to get your help in improving this, I will do the work of development and testing, I only need your guidance.

**Comment by andreiglingeanu** [comment]
Hey @thednp, I am not very familiar with how `FW_Form` helper works. But the proper way would to somehow refactor it to allow that thing. Because the Forms extension [use](https://github.com/ThemeFuse/Unyson-Forms-Extension/blob/master/class-fw-extension-forms.php#L23-L27) this particular helper.

Also, this [`errors_accessed` flag](https://github.com/ThemeFuse/Unyson/blob/d581a4d3a8969f6bc875810de918b66c911fff31/framework/helpers/class-fw-form.php#L390-L393) doesn't look very good too. But I haven't looked very deeply into that thing.

The goal here would be to keep the compatibility with everything else.

**Comment by thednp** [comment]
@andreiglingeanu agreed on goals, however, we cannot continue like this, it's way too "hacky" and not clean (the form errors add a `<style>` to style head for some color styling). This is like Joomla 1.0 to me.

All that not to mention that in 2017/2018 we kinda expect forms to do AJAX validation & submit.

**Comment by andreiglingeanu** [comment]
Well, we are not supposed to be pointing fingers here. We're supposed to come up with ideas and help each other.

Aside from that, what would be the ideal way to deal with form errors, in your vision?

**Comment by thednp** [comment]
Expose the template to views so we can override obviously. And remove the hacky style in the head.

**Comment by andreiglingeanu** [comment]
That would definitely take quite some work. Maybe Viorel could help us here.

**Comment by thednp** [comment]
I propose the following changes regarding the form errors:

**Case 1 - we might not be able to create a template system for errors after all**
* add the ability to set a custom class for the `<ul>` element
* an `apply_filter` would probably do

**Case 2 - the desired and probably expected outcome, but harder to achieve**
* make the form errors have a template system

Now for both cases, we need to disable the `<style>` for form errors AND the ability to set the location where the errors are displayed.

**UPDATE** using AJAX would help solve the problem of styling, so please disregard.

**Comment by thednp** [comment]
For the reviewers and everybody following, I found that enabling AJAX would help solve the problem of styling the form errors, please disregard that. I think we can manage.

Merging the above changes would suffice for now. I am considering adding aria roles to form fields as well, what do you think?

**Comment by thednp** [comment]
So what's up?

**Comment by ViorelEremia** [comment]
@thednp maybe today, is it important to you?

**Comment by thednp** [comment]
@ViorelEremia not exactly, but it's been 2 weeks since the commit and I was thinking you guys aren't interested.

**Comment by ViorelEremia** [comment]
I'm interested, now i'm doing admin notice when are udates for extensions. We plan to solve all the problems from github so I assign them priorities. I'll take a look Monday.

**Comment by ViorelEremia** [comment]
@thednp sorry for delay i have too many users who wait for my answer so i took a quick look on what you did can you inspect why are errors in fronend undefinded layouts something like that. And something is wrong with checkboxes now they are in one line not in new line.

**Comment by thednp** [comment]
@ViorelEremia are you sure you also included [this change](https://github.com/ThemeFuse/Unyson-Forms-Extension/pull/52/commits/01447cf1a75aeaeefa1fe32eddee14d3f6cceaa4)? All changes have to be applied in order for that to work. It works perfect for me.

**Comment by thednp** [comment]
Please do not approve yet, I'm developing some must have functionality with required, aria and other improvements. 

When I'm done, please download my master and copy into your extensions folder of the plugin (rename current installation as a back-up) to do a test.

**Comment by thednp** [comment]
I'm developing a feature to make checkboxes and radios support `required="required"` attribute [properly according to the HTML5 spec](https://stackoverflow.com/questions/5884582/required-attribute-on-multiple-checkboxes-with-the-same-name).

**Comment by thednp** [comment]
I am also researching into browser validation aspects, I hope I can get something to work asap.

@ViorelEremia of @andreiglingeanu what would you prefer for the framework front-end scripting of checkboxes/radios with specific `required` validation scripting?
* a `static/js` file 
* an inline js with that same code?

I would prefer the second, for load performance reasons, we would probably discuss more when I finish the work and commit to my master fork.

**Comment by ViorelEremia** [comment]
You alright with inline js but many developers upload their themes to Envato and reviewers check demos for inline scripts. Try to add it inline later if there will be problems I will add some hooks which will allow manipulating with those scripts.

**Pull request closed without merging**